File: Binary_Heap/decrease_key_min_heap.c
/* File: decrease_key_min_heap.c */

#include <stdio.h>

#define MAX 100

int heap[MAX];
int size = 0;

/* Swap function */
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

/* Insert element */
void insert(int value)
{
    size++;
    int i = size;
    heap[i] = value;

    while(i > 1 && heap[i] < heap[i/2])
    {
        swap(&heap[i], &heap[i/2]);
        i = i/2;
    }
}

/* Decrease Key operation */
void decreaseKey(int i, int newValue)
{
    if(newValue > heap[i])
    {
        printf("New value is greater than current value\n");
        return;
    }

    heap[i] = newValue;

    while(i > 1 && heap[i] < heap[i/2])
    {
        swap(&heap[i], &heap[i/2]);
        i = i/2;
    }
}

/* Print heap */
void printHeap()
{
    for(int i = 1; i <= size; i++)
        printf("%d ", heap[i]);

    printf("\n");
}

int main()
{
    insert(20);
    insert(15);
    insert(30);
    insert(25);
    insert(40);

    printf("Heap before Decrease-Key:\n");
    printHeap();

    decreaseKey(4, 10);

    printf("Heap after Decrease-Key:\n");
    printHeap();

    return 0;
}

File: Binary_Heap/DecreaseKeyMinHeap.java
// File: DecreaseKeyMinHeap.java

class DecreaseKeyMinHeap {

    int[] heap = new int[100];
    int size = 0;

    void swap(int a, int b)
    {
        int temp = heap[a];
        heap[a] = heap[b];
        heap[b] = temp;
    }

    void insert(int value)
    {
        size++;
        int i = size;
        heap[i] = value;

        while(i > 1 && heap[i] < heap[i/2])
        {
            swap(i, i/2);
            i = i/2;
        }
    }

    void decreaseKey(int i, int newValue)
    {
        if(newValue > heap[i])
        {
            System.out.println("New value is greater than current value");
            return;
        }

        heap[i] = newValue;

        while(i > 1 && heap[i] < heap[i/2])
        {
            swap(i, i/2);
            i = i/2;
        }
    }

    void printHeap()
    {
        for(int i = 1; i <= size; i++)
            System.out.print(heap[i] + " ");

        System.out.println();
    }

    public static void main(String[] args)
    {
        DecreaseKeyMinHeap h = new DecreaseKeyMinHeap();

        h.insert(20);
        h.insert(15);
        h.insert(30);
        h.insert(25);
        h.insert(40);

        System.out.println("Heap before Decrease-Key:");
        h.printHeap();

        h.decreaseKey(4, 10);

        System.out.println("Heap after Decrease-Key:");
        h.printHeap();
    }
}
